spacy_experimental/span_boundary_detection/sbd_component.py

```diff
@@ -0,0 +1,301 @@
+            losses during training.
+        threshold (float): Minimum probability to consider a prediction
+            positive.
+        scorer (Optional[Callable]): The scoring method.
+        """
+        self.vocab = nlp
+        self.cfg = {
+            "threshold": threshold,
+        }
+        self.model = model
+        self.name = name
+        self.scorer = scorer
+        Token.set_extension("span_start", default=0, force=True)
+        Token.set_extension("span_end", default=0, force=True)
+
+    def predict(self, docs: Iterable[Doc]):
+        """Apply the pipeline's model to a batch of docs, without modifying them.
+        docs (Iterable[Doc]): The documents to predict.
+        RETURNS: The models prediction for each document.
+        """
+        scores = self.model.predict(docs)
+        return scores
+
+    def set_annotations(self, docs: Iterable[Doc], scores: Floats2d) -> None:
+        """Modify a batch of Doc objects, using pre-computed scores.
+        docs (Iterable[Doc]): The documents to modify.
+        scores: The scores to set, produced by SpanCategorizer.predict.
+        """
+        lengths = [len(doc) for doc in docs]
+
+        offset = 0
+        scores_per_doc = []
+        for length in lengths:
+            scores_per_doc.append(scores[offset : offset + length])
+            offset += length
+
+        for doc, score_doc in zip(docs, scores_per_doc):
+            for token, score_token in zip(doc, score_doc):
+
+                if score_token[0] > self.cfg["threshold"]:
+                    score_token[0] = 1
+                else:
+                    score_token[0] = 0
+
+                if score_token[1] > self.cfg["threshold"]:
+                    score_token[1] = 1
+                else:
+                    score_token[1] = 0
+
+                token._.span_start = score_token[0]
+                token._.span_end = score_token[1]
+
+    def update(
+        self,
+        examples: Iterable[Example],
+        *,
+        drop: float = 0.0,
+        sgd: Optional[Optimizer] = None,
+        losses: Optional[Dict[str, float]] = None,
+    ) -> Dict[str, float]:
+        """Learn from a batch of documents and gold-standard information,
+        updating the pipe's model. Delegates to predict and get_loss.
+        examples (Iterable[Example]): A batch of Example objects.
+        drop (float): The dropout rate.
+        sgd (thinc.api.Optimizer): The optimizer.
+        losses (Dict[str, float]): Optional record of the loss during training.
+            Updated using the component name as the key.
+        RETURNS (Dict[str, float]): The updated losses dictionary.
+        """
+        if losses is None:
+            losses = {}
+        losses.setdefault(self.name, 0.0)
+        references = [eg.reference for eg in examples]
+        set_dropout_rate(self.model, drop)
+        scores, backprop_scores = self.model.begin_update(references)
```

Comment:
I think all other components do the update on the `eg.predicted` https://github.com/explosion/spaCy/blob/master/spacy/pipeline/spancat.py#L339, but I'm not exa tly sure why.

Reply:
It makes sense because `example.predicted` are the docs that don't have the annotations, but since we're only feeding in the raw docs to model it makes no difference for the predictions. I'm using the `example.reference` because I also need them to calculate the score/loss, so I only have to do `references = [eg.reference for eg in examples]` once. But I'm interested in what the other's think? It would certainly look better if I would use `example.predicted`